reps/2022-10-11-serve-java-http-ingress.md

```diff
@@ -0,0 +1,167 @@
+
+Here we use java SPI to implement this feature. We add a `ServeCallableProvider` SPI. 
+```java
+public interface ServeCallableProvider {
+  /**
+   * get Callable type
+   * @return Callable type
+   */
+  String type();
+
+  /**
+   * generate a Callable instance
+   * @param deploymentWrapper deployment info and config
+   * @return Callable instance
+   */
+  Object buildCallable(DeploymentWrapper deploymentWrapper);
+
+  /**
+   * get the signature of callable
+   * @param deploymentWrapper  deployment info and config
+   * @param callable Callable instance
+   * @return
+   */
+  Map<String, Pair<Method, Object>> getSignatures(DeploymentWrapper deploymentWrapper, Object callable);
+}
+```
+The current version of `Callable` is implemented as the default interface implementation. For each additional ingress type, an additional implementation of the `ServeCallableProvider` interface is added.
+
+### method signature cache
+Typically, our `Callable` are user-provided class instances. In JAX-RS, the `Callable` is the jersey application handler when the replica is called using an HTTP proxy. When using the serve handle to call the replica, we cannot call any methods in the class. We are not compatible with serve handle.
+
+On the other hand, every time we call replica, we need to use reflection to get the method that needs to be executed. This will reduce the performance and throughput of the replica
+
+In order to solve the above problems, We need to hold the cache of method signatures.
+
+![image](https://user-images.githubusercontent.com/11265783/195356752-95cf595b-f235-477c-b041-47b3760ad0f5.png)
+
+Ray core uses the signature to decide which method to call. we want to be consistent with it.
+
+When init java replica, we will parse signatures from `Callable` class. Generate the following map. Key is the method signature, value is the pair of the method instance and the `Callable` instance.
+```java
+Map<String, Pair<Method, Object>> signatures;
+```
+If the user configures JAX-rs ingress, we will add one data to the signature cache. The key is always set to `__call__`, the left of the pair in the value is an instance of the `JaxrsIngressCallable.call` method and the right of the pair is an instance of `JaxrsIngressCallable`.
+
+For requests from HTTP proxy, the method signature will be fixed to `__call__`. The request from the serve handle will set the method signature on the client side and hit the signature cache on the server side.
+
+## Compatibility, Deprecation, and Migration Plan
+New features are incremental and do not affect any existing features. And we use SPI to make the modification of the java http ingress meet the open-closed principle.
+## Test Plan and Acceptance Criteria
+- Unit and integration test for core components
+- Benchmarks on java Http ingress performance
```

Comment:
Do we have number to share already? It would be great to show Java HTTP Ingress having good performance for production